Run CI tests in parallel to reduce overall CI time

cderv opened the pull request:
This PR brings lots of changes to how tests are run on CI, but also impact running tests locally and in parallel. 

I tried to document everything in the README (version from this PR here: https://github.com/quarto-dev/quarto-cli/blob/feature/parallel-smokes/tests/README.md) 

## The concept

Idea is to build on `run-parallel-tests.sh` idea to create buckets of tests and then based on that create a Matrix for GHA so that we can start several jobs runner (1 per bucket and per OS) in order to split our tests in GHA in parallel runners. 

At the end, this is working and we decrease overall time even if not perfect. 

## Changes to make that happens 

* `test-smokes-parallel.yml` is now the main CI workflow we trigger. This will run `run-parallel-tests.sh` to get a JSON string for buckets of tests, and use this to create the Matrix that is used to call our `test-smokes.yml` using `workflow_call` event as many time as runner we need. 

* To allow that `run-parallel-tests.ts` has evolved. 
  * It now takes argument, and has a specific flag `--json-for-ci` to create the adequate JSON string. 
  * Challenge was about splitting smoke-all tests documents. So now we can time individual tests and we use that when we create buckets for CI. Otherwise, we put `smoke-all.test.ts` as untimed in its own bucket. The later applies to local run of parallel tests. 

* `test-smokes.yml` behave also differently when run from a `workflow_call` than scheduled tests. The later is still sequential running (one call to `./run-tests.sh`). The former is a for-loop around `run-tests.sh` or `./run-tests.ps1` to run each test in the buckets. 

* Doing that required some changed in `./run-tests.sh`, especially how we run smoke-all tests. 
  * So now you can do `./run-tests.sh docs/smoke-all/2022/**/*.qmd` and the `smoke-all.test.ts` is not needed. (`/run-tests.ps1` can also do that)
  * Timing test will now compute timing for each smoke-all document and record in a file
  * Timing file can now be configured with the env var. 

I may have missed some highlights but as I said 
* It should be documented in README
* There are some comments in the scripts file 
* And code speaks for itself. 


Probably some minor tweaking to do. @cscheid @dragonstyle I would love your review on this. Main concerns

* Does it seems ok to you ? 
* Does `run-tests.sh` still works for you locally on mac ? Same for parallel-tests
* Did I missed anything ? 
* Is there any question ? Which would mean it requires more documentation. 

## Left to do

- [x] I still need to handle new smoke-all test not in the timing file... I forgot to add this....

- [ ] @cscheid as you can see I have now a `timing-for-ci.txt` which detailed tests. I did not overwrite your current timing.txt for now. If the new process fits we can probably have only one file with the detailled timing for both CI and local runs. 

- [ ] Are we ok on the trigger event for the YAML ? Scheduled daily smoke test is still a sequential run as we  don't have timing constraint on this IMO. 

- [ ] Maybe clean the git history of this branch before merging ? Don't know how we usually do it but there are lots of tests commits in there !! Or maybe squash this one ? 

## Later improvmeent

- Add a specific workflow to auto update the timing file directly in CI runners and commit the new timing for our following tests. 


Thanks ! 

(and hopefully this PR won't fail on CI 😅 )

cderv committed:
Adapt buckets creation to handle missing timed tests

cderv committed:
Add CI for buckets creation

cderv committed:
Correct file name

cderv committed:
Speed up by using pre-release version for just using quarto run

cderv committed:
parse argument correctly

cderv committed:
Simplify to check output

cderv committed:
Add inputs to smoke tests

cderv committed:
Try the reusable workflow

cderv committed:
Try the dynamic matrix

cderv committed:
Adapt run-test.ps1 to handle invocation by subprocess

We run the script with `pwsh -F` on CI and we need the arguments to be catched correctly

cderv committed:
For now don't create dependency between jobs

cderv committed:
Output Jobs bucket

cderv committed:
do we format correctly...

cderv committed:
As script and not just command

cderv committed:
Set as outputs

cderv committed:
parse output as JSON

cderv committed:
check matrix is working

cderv committed:
Add a counter for set of buckets

cderv committed:
Try numbering correctly the job for names

cderv committed:
Lets try matrix

cderv committed:
Missing quote

cderv committed:
Missing parenthesis

cderv committed:
Pad the number for better naming

cderv committed:
Correct name for the bucket

cderv committed:
Only add in buckets tests file that still exists

cderv committed:
Don't add test.ts into the timing.txt file

cderv committed:
File is not use in QUARTO_TEST_TIMING value

cderv committed:
Add a echo line to check it runs ok

cderv committed:
Add format to time command

cderv committed:
Update timing

cderv committed:
Uncomment parallel test

cderv committed:
Configure playwright reporter no to open report

cderv committed:
Logging needs to happen after Promise.all()

cderv committed:
QUARTO_TEST_TIMING should be a filename

cderv committed:
Time smoke-all tests individually

cderv committed:
Missing variable use for timing file

cderv committed:
Correctly test for file exception

cderv committed:
correctly test file exception on smoke-all too

cderv committed:
Add a flag to trigger timing in parallel

cderv committed:
Get timing while running locally in parallel

cderv committed:
Change the way we build the files to run in buckets

cderv committed:
store files as a json string

cderv committed:
First try to run command in batch

cderv committed:
Correct quotes

cderv committed:
Fix workflow

cderv committed:
Quotes are needed

cderv committed:
Wrong-copy paste - correct end of foreach

cderv committed:
Add some logging to follow where we are

cderv committed:
Prevent virtual activation on CI for our tests

cderv committed:
`run-tests.sh` now supports a shorter syntax to run smoke-all test documents .qmd and .ipynb

cderv committed:
Rename workflow file

cderv committed:
Run on 20 buckets

Which will be 40 runners (20 x 2 OS)

cderv committed:
Revert "`run-tests.sh` now supports a shorter syntax to run smoke-all test documents .qmd and .ipynb"

This reverts commit 40fd8dc5404eaed79c9ea47620a7449285870106.

cderv committed:
Try 30 buckets

cderv committed:
Revert "Revert "`run-tests.sh` now supports a shorter syntax to run smoke-all test documents .qmd and .ipynb""

This reverts commit 105503eb8d9d738a9b8b114293778b5f28ec0a2c.

cderv committed:
Support short syntax for smoke-all in powershell

cderv committed:
Improve run-tests.sh short syntax for smoke-all

cderv committed:
Format the bucket in TS script directly and use filename only for smoke-all test

cderv committed:
Use 10 buckets for testing

cderv committed:
Add a input to workflow dispatch to better tests

cderv committed:
Fix parallel tests and handle correctly time detailled smoke-all tests (for ci running) and non time detailled smoke-all (for running locally)

cderv committed:
Correct variable to build correct command

cderv committed:
rename and recompute timing file

cderv committed:
Try to handle error in loop for run-test better

cderv committed:
Update the CI for parallel test

cderv committed:
Add some error message to understand what is happening

cderv committed:
Correct -Or operator for Powershell and add color to Error

cderv committed:
Add description to new smoke-test trigger in GHA

cderv committed:
Improve doc on short version to call smoke-all

cderv committed:
Trigger test-smokes-parallel.yml on PR and commit

- Keep scheduled test on non parallel version
- Set default to 20 runners for PR and commit runs

cderv committed:
Add some documentation in test folder README

cderv committed:
Add comments to scripts to document better

cderv committed:
Correctly add untimed smoke-all documents

this insure that new tests will be run even if timing file is outdated

cderv committed:
Error if the timing file does not have timed smoke-all but --json-for-ci is passed

cderv committed:
Ignore `./test.ts` file as it is a helper file only

cderv committed:
Update timing file for ci

cscheid commented on tests/integration/playwright/playwright.config.ts:
```diff
@@ -30,7 +30,7 @@ const config: PlaywrightTestConfig = {
   workers: process.env.CI ? 1 : undefined,
   /* Reporter to use. See https://playwright.dev/docs/test-reporters */
-  reporter: 'html',
+  reporter: [["html", { open: "never" }]],
```
This is by itself a very nice find!

cderv replied:
Thank you ! It was very annoying and in parallel runs locally, I could not have the tests continue running after the browser windows opened and I closed it. So reading mode into playwright doc... and found this ! 😄

cscheid reviewed (commented):
This looks like a great improvement.

My main question here is what happens when we change the number of tests. Specifically, if we add a new test, someone needs to update the timing file. Will it be possible to add a test that is run by the CI if the timing for that file is missing?

In the past, I've been careful to update the timing file myself, but now if the timing file is inconsistent, we need to be totally sure that we still run the new test.

cderv commented:
@cscheid 

Very good question and feedback ! I realized that also yesterday, and I definitely need to add a trick for non-timed `smoke-all` documents, like you did for other smoke test (adding them with 0 timing in a random bucket). I'll add that! 

Also I was thinking of having a the timing file auto update using another GHA workflow - maybe weekly and by opening a PR so that we can check if it is ok before merging the new timed file.

cderv commented:
I added new commits that should solve the issue of untimed tests. Everything should be assigned to a bucket now. 

I am seeing that my commit have the wrong time (WSL linux issue it seems 🤦‍♂️ ) - first time I am working to push to Quarto using Linux. that is why the commits are appearing before your comment, even if I have just made those today

My linux environment has this date 😠 
````
$ date
Wed Jul 19 11:20:41 CEST 2023
````

Crazy bug on WSL Ubuntu 20 !! I need to update this... 

Hopefully it is not too important to finish the work.

cderv commented:
> I added new commits that should solve the issue of untimed tests. Everything should be assigned to a bucket now.

I confirmed this is working - some untimed test were correctly added. I have now updated the timing file and rebased everything. We should be good to go !

cderv commented:
So it's live ! Hopefully, all we be going well. We'll see! I'll watch out
